File: src/Java/myservicebus/src/main/java/com/myservicebus/topology/EndpointDefinitionModel.java
package com.myservicebus.topology;

/** Immutable endpoint policy captured during registration normalization. */
public record EndpointDefinitionModel(
        String name,
        boolean nameExplicit,
        Class<?> nameFormatterType,
        Integer concurrentMessageLimit,
        Integer prefetchCount) {
    public EndpointDefinitionModel {
        if (name == null || name.isBlank()) {
            throw new IllegalArgumentException("Endpoint name must not be blank.");
        }
        if (concurrentMessageLimit != null && concurrentMessageLimit < 1) {
            throw new IllegalArgumentException("Concurrent message limit must be greater than zero.");
        }
        if (prefetchCount != null && prefetchCount < 1) {
            throw new IllegalArgumentException("Prefetch count must be greater than zero.");
        }
    }
}

File: src/MyServiceBus/Topology/EndpointDefinitionModel.cs
namespace MyServiceBus.Topology;

/// <summary>
/// Immutable endpoint policy captured during registration normalization.
/// </summary>
public sealed record EndpointDefinitionModel
{
    public EndpointDefinitionModel(
        string name,
        bool nameIsExplicit,
        Type? nameFormatterType,
        int? concurrentMessageLimit,
        ushort? prefetchCount)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(name);
        if (concurrentMessageLimit is <= 0)
            throw new ArgumentOutOfRangeException(nameof(concurrentMessageLimit));
        if (prefetchCount is 0)
            throw new ArgumentOutOfRangeException(nameof(prefetchCount));

        Name = name;
        NameIsExplicit = nameIsExplicit;
        NameFormatterType = nameFormatterType;
        ConcurrentMessageLimit = concurrentMessageLimit;
        PrefetchCount = prefetchCount;
    }

    public string Name { get; }

    public bool NameIsExplicit { get; }

    public Type? NameFormatterType { get; }

    public int? ConcurrentMessageLimit { get; }

    public ushort? PrefetchCount { get; }
}
